File: backend/src/main/java/com/poultry/dispute/entity/Dispute.java
package com.poultry.dispute.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "disputes")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Dispute
{

  public enum RaisedBy
  {
    BUYER,
    SELLER
  }

  public enum DisputeType
  {
    QUALITY_ISSUE,
    QUANTITY_MISMATCH,
    DELIVERY_ISSUE,
    PAYMENT_ISSUE,
    WRONG_PRODUCT,
    DAMAGED_GOODS,
    OTHER
  }

  public enum DisputeStatus
  {
    OPEN,
    UNDER_REVIEW,
    AWAITING_RESPONSE,
    RESOLVED,
    ESCALATED,
    CLOSED
  }

  public enum DisputeResolution
  {
    REFUND_FULL,
    REFUND_PARTIAL,
    REPLACEMENT,
    CREDIT_NOTE,
    NO_ACTION,
    MUTUAL_AGREEMENT
  }

  public boolean isOpen()
  {
    return status == DisputeStatus.OPEN || status == DisputeStatus.UNDER_REVIEW
        || status == DisputeStatus.AWAITING_RESPONSE || status == DisputeStatus.ESCALATED;
  }

  public boolean canBeEscalated()
  {
    return status == DisputeStatus.OPEN || status == DisputeStatus.UNDER_REVIEW
        || status == DisputeStatus.AWAITING_RESPONSE;
  }

  public void addMessage(DisputeMessage message)
  {
    messages.add(message);
    message.setDispute(this);
  }
  @Id
  @GeneratedValue(strategy = GenerationType.UUID)
  private UUID id;
  @Column(name = "dispute_number", nullable = false, unique = true)
  private String disputeNumber;
  @Column(name = "order_id", nullable = false)
  private UUID orderId;
  @Column(name = "buyer_id", nullable = false)
  private UUID buyerId;
  @Column(name = "seller_id", nullable = false)
  private UUID sellerId;
  @Enumerated(EnumType.STRING)
  @Column(name = "raised_by", nullable = false, columnDefinition = "dispute_raised_by")
  private RaisedBy raisedBy;
  @Enumerated(EnumType.STRING)
  @Column(nullable = false, columnDefinition = "dispute_type")
  private DisputeType type;
  @Enumerated(EnumType.STRING)
  @Column(nullable = false, columnDefinition = "dispute_status")
  @Builder.Default
  private DisputeStatus status = DisputeStatus.OPEN;
  @Column(nullable = false, length = 200)
  private String title;
  @Column(nullable = false, columnDefinition = "TEXT")
  private String description;
  @JdbcTypeCode(SqlTypes.JSON)
  @Column(name = "evidence_urls", columnDefinition = "jsonb")
  @Builder.Default
  private List<String> evidenceUrls = new ArrayList<>();
  @Enumerated(EnumType.STRING)
  @Column(name = "requested_resolution", columnDefinition = "dispute_resolution")
  private DisputeResolution requestedResolution;
  @Column(name = "requested_amount", precision = 12, scale = 2)
  private BigDecimal requestedAmount;
  @Enumerated(EnumType.STRING)
  @Column(name = "final_resolution", columnDefinition = "dispute_resolution")
  private DisputeResolution finalResolution;
  @Column(name = "resolution_amount", precision = 12, scale = 2)
  private BigDecimal resolutionAmount;
  @Column(name = "resolution_notes", columnDefinition = "TEXT")
  private String resolutionNotes;
  @Column(name = "assigned_to")
  private UUID assignedTo;
  @Column(nullable = false)
  @Builder.Default
  private Integer priority = 2;
  @Column(name = "escalated_at")
  private Instant escalatedAt;
  @Column(name = "resolved_at")
  private Instant resolvedAt;
  @Column(name = "resolved_by")
  private UUID resolvedBy;
  @OneToMany(mappedBy = "dispute", cascade = CascadeType.ALL, orphanRemoval = true)
  @Builder.Default
  private List<DisputeMessage> messages = new ArrayList<>();
  @CreationTimestamp
  @Column(name = "created_at", nullable = false, updatable = false)
  private Instant createdAt;
  @UpdateTimestamp
  @Column(name = "updated_at", nullable = false)
  private Instant updatedAt;
}

File: seller-web/src/types/dispute.ts
// Dispute Types
export interface Dispute {
  id: string;
  disputeNumber: string;
  orderId: string;
  orderNumber: string;
  buyerId: string;
  buyerName: string;
  sellerId: string;
  type: DisputeType;
  status: DisputeStatus;
  priority: DisputePriority;
  reason: string;
  description: string;
  requestedResolution: string;
  resolution?: string;
  refundAmount?: number;
  attachments: DisputeAttachment[];
  messages: DisputeMessage[];
  createdAt: string;
  updatedAt: string;
  resolvedAt?: string;
  resolvedBy?: string;
}

export type DisputeType =
  | "QUALITY_ISSUE"
  | "QUANTITY_MISMATCH"
  | "WRONG_PRODUCT"
  | "DAMAGED_PRODUCT"
  | "LATE_DELIVERY"
  | "NON_DELIVERY"
  | "PAYMENT_ISSUE"
  | "OTHER";

export type DisputeStatus =
  | "OPEN"
  | "UNDER_REVIEW"
  | "WAITING_SELLER_RESPONSE"
  | "WAITING_BUYER_RESPONSE"
  | "ESCALATED"
  | "RESOLVED"
  | "CLOSED";

export type DisputePriority = "LOW" | "MEDIUM" | "HIGH" | "URGENT";

export interface DisputeAttachment {
  id: string;
  fileName: string;
  fileUrl: string;
  fileType: string;
  fileSize: number;
  uploadedAt: string;
  uploadedBy: string;
}

export interface DisputeMessage {
  id: string;
  disputeId: string;
  senderId: string;
  senderName: string;
  senderType: "BUYER" | "SELLER" | "ADMIN";
  message: string;
  attachments: DisputeAttachment[];
  createdAt: string;
  isInternal: boolean;
}

export interface CreateDisputeResponseRequest {
  message: string;
  proposedResolution?: string;
  refundAmount?: number;
  attachments?: File[];
}

export interface DisputeFilters {
  status?: DisputeStatus | "ALL";
  type?: DisputeType | "ALL";
  priority?: DisputePriority | "ALL";
  page?: number;
  size?: number;
}
